src/main/java/com/aws/greengrass/util/platforms/unix/UnixPlatform.java

```diff
@@ -376,76 +369,73 @@ public void addUserToGroup(String user, String group) throws IOException {
     }
 
     @Override
-    public void setPermissions(FileSystemPermission permission, Path path, EnumSet<Option> options)
-            throws IOException {
+    protected void setOwner(FileSystemPermission permission, Path path) throws IOException {
+        UserPrincipalLookupService lookupService = path.getFileSystem().getUserPrincipalLookupService();
+        UserPrincipal userPrincipal = lookupService.lookupPrincipalByName(permission.getOwnerUser());
+        GroupPrincipal groupPrincipal = Utils.isEmpty(permission.getOwnerGroup()) ? null :
+                lookupService.lookupPrincipalByGroupName(permission.getOwnerGroup());
+        PosixFileAttributeView view = Files.getFileAttributeView(path, PosixFileAttributeView.class,
+                LinkOption.NOFOLLOW_LINKS);
 
-        // noop function that does not set owner
-        CrashableFunction<PosixFileAttributeView, Void, IOException> setOwner = (p) -> null;
+        logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path).kv("owner", permission.getOwnerUser())
+                .log();
+        view.setOwner(userPrincipal);
 
-        if (options.contains(Option.SetOwner)) {
-            if (Utils.isEmpty(permission.getOwnerUser())) {
-                logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path).log("No owner to set for path");
-            } else {
-                UserPrincipalLookupService lookupService = path.getFileSystem().getUserPrincipalLookupService();
-                UserPrincipal userPrincipal = lookupService.lookupPrincipalByName(permission.getOwnerUser());
-                GroupPrincipal groupPrincipal = Utils.isEmpty(permission.getOwnerGroup()) ? null :
-                        lookupService.lookupPrincipalByGroupName(permission.getOwnerGroup());
-
-                setOwner = (view) -> {
-                    logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path)
-                            .kv("owner", permission.getOwnerUser()).log();
-                    view.setOwner(userPrincipal);
-                    if (groupPrincipal != null) {
-                        logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path)
-                                .kv("group", permission.getOwnerGroup()).log();
-                        view.setGroup(groupPrincipal);
-                    }
-                    return null;
-                };
-            }
+        if (groupPrincipal != null) {
+            logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path).kv("group",
+                    permission.getOwnerGroup()).log();
+            view.setGroup(groupPrincipal);
         }
+    }
 
+    @Override
+    protected void setMode(FileSystemPermission permission, Path path) throws IOException {
+        Set<PosixFilePermission> perms = posixFilePermissions(permission);
```

Comment:
similar problem as with the owner - we now have to convert the permissions object for every file and allocate a new set everytime.

Throwing this out as a suggestion:

We can do something like have a `FileSystemPermissionView` abstract class which has concrete `Posix` and `Windows` implementations where we can precalculate any platform specific stuff.

Have a platform method like `Platform.getFileSystemPermissionView(FileSystemPermission perm)`

The `posixFilePermissions` method you have would just be a method on the `PosixFileSystemPermissionView`
You could calculate the permissions once when constructing the object and pass the `FileSystemPermissionView` into `setMode`, cast it as a the Posix version and then use those permissions.